pipeline_dp/pld_budget_accounting.py

```diff
@@ -0,0 +1,114 @@
+"""Privacy Loss Distribution (PLD) budget accounting for DP Pipelines."""
+
+from dataclasses import dataclass, field
+from pipeline_dp.aggregate_params import NoiseKind
+from dp_accounting import privacy_loss_distribution
+
+
+@dataclass
+class PLDBudget:
+    _noise_parameter: float = None  # i am not sure how to use this, as noise is calibrated globally?
+
+    @property
+    def noise_parameter(self):
+        if self._noise_parameter is None:
+            raise AssertionError("Noise parameter is not yet defined.")
+        return self._noise_parameter
+
+
+@dataclass
+class RequestedPLDBudget:
+    budget: PLDBudget
+
+
+@dataclass
+class PLDBudgetAccountant:
+    noise: float = 1  # this is where i am storing noise, does this make sense?
+    max_noise: float = 50
+    sensitivity: float = 1
+    total_epsilon: float = None
+    total_delta: float = None
+    _requested_laplace_budgets: list = field(default_factory=list)
+    _requested_gaussian_budgets: list = field(default_factory=list)
+
+    def __post_init__(self):
+        if self.total_epsilon is None or self.total_delta is None:
+            raise ValueError(f"Must pass total_epsilon and total_delta")
+        if self.total_epsilon is not None and self.total_epsilon <= 0:
+            raise ValueError(
+                f"Epsilon must be positive, not {self.total_epsilon}.")
+        if self.total_delta is not None and self.total_delta < 0:
+            raise ValueError(
+                f"Delta must be non-negative, not {self.total_delta}.")
+        if self.noise > self.max_noise:
+            raise ValueError(
+                f"Noise = {self.noise} is larger than the maximum = {self.max_noise}."
+            )
+
+    def request_budget(self, noise_kind: NoiseKind) -> PLDBudget:
+        pldBudget = PLDBudget()
+        if noise_kind == NoiseKind.LAPLACE:
+            self._requested_laplace_budgets.append(
+                RequestedPLDBudget(pldBudget))
+        elif noise_kind == NoiseKind.GAUSSIAN:
+            self._requested_gaussian_budgets.append(
+                RequestedPLDBudget(pldBudget))
+        else:
+            raise NotImplementedError(
+                f"Noise type {noise_kind} is unsupported.")
+        return pldBudget
+
+    def compute_budgets(self):
+        optimal_noise = self.optimize_noise_to_delta()
+        self.noise = optimal_noise  # i set the noise for the entire budget accountant
+
+    def optimize_noise_to_delta(self) -> float:
+        """Adjust noise to get close to given epsilon; increasing noise decreases epsilon."""
+        threshold = 0.001
+        optimal_noise = self.noise
+        pld_laplace_gaussian = self.compose_distributions(optimal_noise)
+        pld_epsilon = pld_laplace_gaussian.get_epsilon_for_delta(
+            self.total_delta)
+        (min, max) = (0, self.max_noise)
```

Comment:
Nit: remove ( )
Nit: replace min/max, to something like low/high (or other similar names). 

That's because min/max are built-in functions, it's better not to hide built-in functions.

Reply:
👍  Changed, thanks!